src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializerHelpers.cs

```diff
@@ -0,0 +1,364 @@
+﻿//-----------------------------------------------------------------------------
+// <copyright file="ODataResourceDeserializerHelpers.cs" company=".NET Foundation">
+//      Copyright (c) .NET Foundation and Contributors. All rights reserved. 
+//      See License.txt in the project root for license information.
+// </copyright>
+//------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Linq;
+using System.Reflection;
+using Microsoft.AspNet.OData.Interfaces;
+using Microsoft.OData;
+using Microsoft.OData.Edm;
+using Microsoft.OData.UriParser;
+
+namespace Microsoft.AspNet.OData.Formatter.Deserialization
+{
+    internal static class ODataResourceDeserializerHelpers
+    {
+        internal static void AppendDeletedProperties(dynamic resource, ODataDeletedResource deletedResource, bool isUntyped)
+        {
+            if (isUntyped)
+            {
+                resource.Id = deletedResource.Id.ToString();
+            }
+            else
+            {
+                resource.Id = deletedResource.Id;
+            }
+
+            if (deletedResource.Reason != null)
+            {
+                resource.Reason = deletedResource.Reason.Value;
+            }
+        }
+
+        /// <summary>
+        /// Creates a nested read context for nested resources.
+        /// It ensures that the read context and the path are always correct for the level that is being read.
+        /// </summary>
+        /// <param name="resourceInfoWrapper">The nested resource info wrapper.</param>
+        /// <param name="readContext">The read context.</param>
+        /// <param name="edmProperty">The Edm property.</param>
+        /// <returns>It returns the nested <see cref="ODataDeserializerContext"/> for the resource being read.</returns>
+        internal static ODataDeserializerContext GenerateNestedReadContext(ODataNestedResourceInfoWrapper resourceInfoWrapper, ODataDeserializerContext readContext, IEdmProperty edmProperty)
+        {
+            ODataDeserializerContext nestedReadContext;
+            Routing.ODataPath path = readContext.Path;
```

Comment:
nit: I wish we could come up with a _disambiguating_ alias for `Routing.ODataPath` because most people may not know the reason we're using that is because the type exists in both ODL and WebApi libraries.
But if we can't come up with a good one then it's fine to leave it as is.